### PR TITLE
add ForwardDiff and ChainRules extension

### PR DESCRIPTION
**Pull request by longemen3000**
addresses #28, eq 18 of https://mathworld.wolfram.com/Polylogarithm.html , only real numbers

In theory, the same relation can be done to support the complex domain, but ForwardDiff is a little bit tricky with complex numbers. at least the skeleton (extension loading) is done.

this package supports Julia versions >=1.0, but extensions are only supported in Julia >= 1.9. that means that AD support is only available to new Julia versions. (one alternative would be to add Requires.jl, but it is more cumbersome)

The only thing missing is ChainRules test. (that could be done by loading Zygote, for example, or using ChainRulesTestUtils)

### CODE REVIEW

File: ext/PolyLogChainRulesExt.jl [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module PolyLogChainRulesExt
+
+using PolyLog
+using ChainRulesCore
+@scalar_rule reli4(x) reli3(x)/x
+@scalar_rule reli3(x) reli2(x)/x
+@scalar_rule reli2(x) reli1(x)/x
+@scalar_rule reli1(x) one(x)/(one(x)-x)
+@scalar_rule reli(n,x) reli(n-1,x)/x
+end #module
```
Comment (Expander) [inline ext/PolyLogChainRulesExt.jl:10]: I see a warning that this file does not end with a newline.

File: ext/PolyLogForwardDiffExt.jl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+end
+
+function PolyLog.reli(n::Integer,d::Dual{T}) where T
+    val = ForwardDiff.value(d)
+    x = reli(val)
+    dx = PolyLog.reli(n-1,val)/val
+    return Dual{T}(x, dx*partials(d))
+end
+
+end #module
```
Comment (Expander) [inline ext/PolyLogForwardDiffExt.jl:41]: I see another warning that this file does not end with a newline, too.

**Comment by Expander** [comment]
Thanks for this PR!

When I
~~~.jl
julia --project test/runtests.jl
~~~
I get the following error:
~~~.jl
Test Summary: | Pass  Total  Time
digamma       |   11     11  0.1s
Test Summary: | Pass  Total  Time
eta           |   28     28  0.3s
Test Summary: | Pass  Total  Time
fac           |   19     19  0.1s
Test Summary: | Pass  Total  Time
inv_fac       |   20     20  0.1s
Test Summary: | Pass  Total  Time
harmonic      |   19     19  0.2s
Test Summary: |  Pass  Total  Time
li0           | 41267  41267  2.8s
Test Summary: |  Pass  Total  Time
li1           | 41397  41397  2.3s
li2: Error During Test at /home/avoigt/research/PolyLog.jl/test/Li2.jl:71
  Test threw exception
  Expression: ForwardDiff.derivative(PolyLog.reli2, float(pi)) == reli1(float(pi)) / float(pi)
  UndefVarError: `reli1` not defined
  Stacktrace:
   [1] macro expansion
     @ ~/.julia/juliaup/julia-1.10.3+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:669 [inlined]
   [2] macro expansion
     @ ~/research/PolyLog.jl/test/Li2.jl:71 [inlined]
   [3] macro expansion
     @ ~/.julia/juliaup/julia-1.10.3+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [4] top-level scope
     @ ~/research/PolyLog.jl/test/Li2.jl:2
Test Summary: |   Pass  Error   Total   Time
li2           | 102820      1  102821  22.2s
ERROR: LoadError: Some tests did not pass: 102820 passed, 0 failed, 1 errored, 0 broken.
in expression starting at /home/avoigt/research/PolyLog.jl/test/Li2.jl:1
in expression starting at /home/avoigt/research/PolyLog.jl/test/runtests.jl:14
~~~

**Comment by Expander** [comment]
Sorry, I still get the same error. I think you should replace `ForwardDiff.reli` by `PolyLog.reli` etc.

**Comment by longemen3000** [comment]
> Sorry, I still get the same error. I think you should replace `ForwardDiff.reli` by `PolyLog.reli` etc.

uff, good catch!

**Comment by Expander** [comment]
Thanks for fixing this!

However, I still get another error:
~~~.jl
Test Summary: |  Pass  Total  Time
li6           | 69334  69334  0.8s
li: Error During Test at /home/avoigt/research/PolyLog.jl/test/Li.jl:133
  Test threw exception
  Expression: ForwardDiff.derivative(Base.Fix1(PolyLog.reli, 6), float(pi)) == PolyLog.reli(5, float(pi)) / float(pi)
  MethodError: no method matching reli(::Float64)
  
  Closest candidates are:
    reli(::Integer, ::ForwardDiff.Dual{T}) where T
     @ PolyLogForwardDiffExt ~/research/PolyLog.jl/ext/PolyLogForwardDiffExt.jl:34
    reli(::Integer, ::Real)
     @ PolyLog ~/research/PolyLog.jl/src/Li.jl:26
  
  Stacktrace:
   [1] reli(n::Int64, d::ForwardDiff.Dual{ForwardDiff.Tag{Base.Fix1{typeof(PolyLog.reli), Int64}, Float64}, Float64, 1})
     @ PolyLogForwardDiffExt ~/research/PolyLog.jl/ext/PolyLogForwardDiffExt.jl:36
   [2] (::Base.Fix1{typeof(PolyLog.reli), Int64})(y::ForwardDiff.Dual{ForwardDiff.Tag{Base.Fix1{typeof(PolyLog.reli), Int64}, Float64}, Float64, 1})
     @ Base ./operators.jl:1118
   [3] derivative(f::Base.Fix1{typeof(PolyLog.reli), Int64}, x::Float64)
     @ ForwardDiff ~/.julia/packages/ForwardDiff/PcZ48/src/derivative.jl:14
   [4] macro expansion
     @ ~/.julia/juliaup/julia-1.10.3+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:669 [inlined]
   [5] macro expansion
     @ ~/research/PolyLog.jl/test/Li.jl:133 [inlined]
   [6] macro expansion
     @ ~/.julia/juliaup/julia-1.10.3+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [7] top-level scope
     @ ~/research/PolyLog.jl/test/Li.jl:7
Test Summary: |    Pass  Error    Total     Time
li            | 3038838      1  3038839  3m15.3s
ERROR: LoadError: Some tests did not pass: 3038838 passed, 0 failed, 1 errored, 0 broken.
~~~
To fix it, it would be best to extend the test to "arbitrary" `n` (i.e. not only $n=6$) by moving the test into the for loop `for ni in nis` and test `reli(n,x)` against `reli(n-1,x)/x`.

**Comment by Expander** [comment]
Thanks! Please also move the comment `#ForwardDiff Test` to the loop. :)

**Comment by Expander** [comment]
Sorry to be pedantic, but could you please also add a test for the derivative of $Li_n(x)$ at $x=0$? If I'm not mistaken, we should have $Li_{n}'(0) = 1$.

**Comment by longemen3000** [comment]
hmm, good insight, the current definitions don't account for that. but it is just the matter of a branch in the derivative definition

**Comment by Expander** [comment]
Could you please merge the `main` branch into your branch that you're using for this PR? (It contains a change so we can run all github actions, including the build and test.)

**Comment by longemen3000** [comment]
Is it possible to move the minimum Julia version to 1.6? (the last long term support release)

**Comment by Expander** [comment]
I get another error from the github actions:
~~~.jl
li: Test Failed at /home/runner/work/PolyLog.jl/PolyLog.jl/test/Li.jl:98
  Expression: ForwardDiff.derivative(Base.Fix1(PolyLog.reli, n), float(pi)) == PolyLog.reli(n, float(pi)) / float(pi)
   Evaluated: 2.509398245036805e6 == -261143.92161803515
Stacktrace:
 [1] macro expansion
   @ /opt/hostedtoolcache/julia/1.10.3/x64/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] macro expansion
   @ ~/work/PolyLog.jl/PolyLog.jl/test/Li.jl:98 [inlined]
 [3] macro expansion
   @ /opt/hostedtoolcache/julia/1.10.3/x64/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [4] top-level scope
   @ ~/work/PolyLog.jl/PolyLog.jl/test/Li.jl:7
~~~
It seems to me that the test is incorrect: It seems the test checks that $Li_n(x)=Li_n(x)/x$. How can that be?

**Comment by Expander** [comment]
> Is it possible to move the minimum Julia version to 1.6? (the last long term support release)

I guess you're asking because of the failing test for julia 1.0?

Just for my understanding: It would require all dependents to also upgrade to 1.6 right? And it would require to move to version 3.0.0, I guess. We would also give up navite support for Debian's `oldstable` (bullseye).

**Comment by longemen3000** [comment]
Ohh, wait, I know why this is failing, one moment

File: test/Li.jl [hunk 3]
```diff
@@ -93,6 +93,12 @@ end
         @test PolyLog.li(n, 1//1 + 0//1im) ≈ zeta
         @test PolyLog.li(n, 1 + 0im) ≈ zeta
         @test PolyLog.li(n, BigFloat("1.0") + 0im) == PolyLog.zeta(n, BigFloat)
+
+        #ForwardDiff Test
+        if isdefined(Base,:get_extension)
```
Comment (Expander) [inline test/Li.jl:98]: Could you please check if all these `if isdefined(...)` statements are still necessary after your recent commits? If not, could you please remove them?

**Comment by Expander** [comment]
Great, now all tests pass!

I've noticed that coveralls shows some code that has not been tested, see attached plot. Would you mind checking if it is covered?
coveralls

**Comment by Expander** [comment]
Thanks for fixing this!
Are these `if isdefined(Base, :get_extension)` statements still necessary for the compatibility with julia 1.0?

**Comment by longemen3000** [comment]
Yeah, the ForwardDiff support is only available after 1.9.
You could support this in 1.0, but that would require to add additional dependencies (`Requires.jl`)

**Comment by Expander** [comment]
Ok, thanks for the answer!

**Comment by Expander** [comment]
I think so far the chain rules are untested, right? You mentioned that this would be possible with the `ChainRulesTestUtils`. Would it be possible to add such tests?

File: test/Li.jl [hunk 4]
```diff
@@ -127,5 +135,5 @@ end
                 @test typeof(PolyLog.reli(n, convert(T, x))) == T
             end
         end
-    end
+    end
```
Comment (Expander) [inline test/Li.jl:138]: Sorry to be pedantic, but would you mind removing this trailing whitespace? :)

Comment (longemen3000) [reply test/Li.jl:138]: no problem!

Comment (Expander) [reply test/Li.jl]: Thanks! :)

**Comment by Expander** [comment]
First of all, thanks for adding the tests!

One last thing: I think we also need appropriate entries in the `[compat]` section for `ForwardDiff` and `ChainRulesCore`, right?. Would you mind adding these? (I think it would be best to have a `[compat]` entry that allows for the most wide range of compatible versions.)

**Comment by Expander** [comment]
I currently get the following error:
~~~.jl
┌ Warning: Could not use exact versions of packages in manifest, re-resolving
└ @ Pkg.Operations /opt/hostedtoolcache/julia/1.10.4/x64/share/julia/stdlib/v1.10/Pkg/src/Operations.jl:1814
ERROR: LoadError: Unsatisfiable requirements detected for package ForwardDiff [f6369f11]:
 ForwardDiff [f6369f11] log:
 ├─possible versions are: 0.9.0-0.10.36 or uninstalled
 ├─restricted to versions 0.11 by PolyLog [85e3b03c], leaving only versions: uninstalled
 │ └─PolyLog [85e3b03c] log:
 │   ├─possible versions are: 2.4.2 or uninstalled
 │   └─PolyLog [85e3b03c] is fixed to version 2.4.2
 └─restricted to versions 0.11 by an explicit requirement — no versions left
~~~

**Comment by Expander** [comment]
Excellent, thanks a lot! If it is ok with you, I'll merge this PR.

Are you ok with that I add you to a list of contributors?

One final question: Would you mind telling me when version 1.0 of `ForwardDiff` is out?

**Comment by longemen3000** [comment]
Thanks!, I don't mind, do want you feel is best

On ForwardDiff, No idea, it has been years

**Pull request merged**